source/logrepl/snapshot.go

```diff
@@ -0,0 +1,213 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logrepl
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/conduitio/conduit-connector-postgres/pgutil"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/jackc/pgtype"
+	"github.com/jackc/pgx/v4"
+)
+
+var psql = sq.StatementBuilder.PlaceholderFormat(sq.Dollar)
+
+const actionSnapshot string = "snapshot"
+
+type SnapshotConfig struct {
+	SnapshotName string
+	Table        string
+	Columns      []string
+	KeyColumn    string
+}
+
+type SnapshotIterator struct {
+	config SnapshotConfig
+
+	tx   pgx.Tx
+	rows pgx.Rows
+
+	internalPos int64
+}
+
+func NewSnapshotIterator(ctx context.Context, conn *pgx.Conn, cfg SnapshotConfig) (*SnapshotIterator, error) {
+	s := &SnapshotIterator{
+		config: cfg,
+	}
+
+	err := s.startSnapshotTx(ctx, conn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start snapshot tx: %w", err)
+	}
+
+	err = s.loadRows(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load rows: %w", err)
+	}
+
+	return s, nil
+}
+
+func (s *SnapshotIterator) loadRows(ctx context.Context) error {
+	query, args, err := psql.
+		Select(s.config.Columns...).
+		From(s.config.Table).
+		ToSql()
+	if err != nil {
+		s.rows.Close()
+		return fmt.Errorf("failed to create read query: %w", err)
+	}
+
+	rows, err := s.tx.Query(ctx, query, args...)
+	if err != nil {
+		s.rows.Close()
+		return fmt.Errorf("failed to query rows: %w", err)
+	}
+	s.rows = rows
+
+	return nil
+}
+
+func (s *SnapshotIterator) startSnapshotTx(ctx context.Context, conn *pgx.Conn) error {
+	tx, err := conn.BeginTx(ctx, pgx.TxOptions{
+		IsoLevel:   pgx.RepeatableRead,
+		AccessMode: pgx.ReadOnly,
+	})
+	if err != nil {
+		return err
+	}
+
+	s.tx = tx
+
+	snapshotTx := fmt.Sprintf(`SET TRANSACTION SNAPSHOT '%s'`, s.config.SnapshotName)
+	_, err = tx.Exec(ctx, snapshotTx)
+	if err != nil {
+		if rollErr := s.tx.Rollback(ctx); rollErr != nil {
+			sdk.Logger(ctx).Err(rollErr).Msg("set transaction rollback failed")
+		}
+		return fmt.Errorf("failed to set transaction snapshot id: %w", err)
+	}
+
+	return nil
+}
+
+func (s *SnapshotIterator) Next(ctx context.Context) (sdk.Record, error) {
```

Comment:
One more important thing I forgot last time - if the connector receives a stop signal this context will get cancelled but it will continue calling the`Read` method to give the connector a chance to empty its cache. Long story short - we need to check if the context is cancelled and return `ctx.Err()` if it is, to ensure the connector can stop in the middle of a snapshot.